src/txn/txn.c

```diff
@@ -1257,15 +1258,20 @@ __txn_commit_timestamps_assert(WT_SESSION_IMPL *session)
     wt_timestamp_t prev_op_durable_ts, prev_op_ts;
     u_int i;
     bool op_zero_ts, upd_zero_ts, used_ts;
+    bool recovery_conn;
 
     txn = session->txn;
     cursor = NULL;
+    conn = S2C(session);
 
     used_ts = F_ISSET(txn, WT_TXN_HAS_TS_COMMIT) || F_ISSET(txn, WT_TXN_HAS_TS_DURABLE);
+    recovery_conn = F_ISSET(conn, WT_CONN_RECOVERING);
     /*
-     * Debugging checks on timestamps, if user requested them.
+     * Debugging checks on timestamps, if user requested them. We additionally don't expect recovery
+     * to be using timestamps when applying commits. If recovery is running, skip this assert to
+     * avoid failing the recovery process.
      */
-    if (F_ISSET(txn, WT_TXN_TS_WRITE_ALWAYS) && !used_ts && txn->mod_count != 0)
+    if (F_ISSET(txn, WT_TXN_TS_WRITE_ALWAYS) && !used_ts && txn->mod_count != 0 && !recovery_conn)
```

Comment:
You're only using both local variables once. I would simplify it by removing those and making the conditional as 
`!F_ISSET(S2C(session), WT_CONN_RECOVERING)`.

Reply:
Good point! I've removed the local variable and added `!F_ISSET(S2C(session), WT_CONN_RECOVERING)` into the conditional.